### PR TITLE
SAA-716: new patch endpoint to update allocations

### PR DESCRIPTION
**Pull request by gerryallan-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt [hunk 1]
```diff
@@ -1,18 +1,64 @@
...
+
+    return allocation.toModel()
+  }
+
+  private fun applyStartDateUpdate(
+    request: AllocationUpdateRequest,
+    allocation: Allocation,
+  ) {
+    request.startDate?.apply {
+      allocation.startDate = this
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt:43]: We are going to need checks around changing the start and end dates.  We can't just change them.

For example the start and end dates cannot go beyond that of the activity or schedule.

Comment (gerryallan-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt:50]: Added validation for start and end date

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt [hunk 2]
```diff
@@ -1,18 +1,86 @@
...
     allocationRepository
       .findByPrisonCodeAndPrisonerNumbers(prisonCode, prisonNumbers.toList())
       .filter { !activeOnly || it.status(PrisonerStatus.ACTIVE) }
       .toModelPrisonerAllocations()
 
   fun getAllocationById(id: Long) = allocationRepository.findOrThrowNotFound(id).toModel()
+
+  @PreAuthorize("hasAnyRole('ACTIVITY_HUB', 'ACTIVITY_HUB_LEAD', 'ACTIVITY_ADMIN')")
+  fun updateAllocation(allocationId: Long, request: AllocationUpdateRequest, prisonCode: String): ModelAllocation {
+    var allocation = allocationRepository.findOrThrowNotFound(allocationId)
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt:28]: @timharrison-moj 

I think we need to guard against updating allocations that are ended.

We may want to consider other states but should check not ended as a minimum.

```
if (allocation.status(PrisonerStatus.ENDED)) {
  throw IllegalStateException(....)
}
```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt [hunk 3]
```diff
@@ -1,18 +1,86 @@
...
+    }
+  }
+
+  private fun applyRemoveEndDateUpdate(
+    request: AllocationUpdateRequest,
+    allocation: Allocation,
+  ) {
+    request.removeEndDate?.apply {
+      if (this) {
+        allocation.endDate = null
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt:73]: @timharrison-moj 

End date can be derived.  What I mean by that is itmight not be set directly but come from a planned deallocation.

So should this removed the planned deallocation if there is one?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/Allocation.kt [hunk 4]
```diff
@@ -28,15 +28,15 @@ data class Allocation(
   val isUnemployment: Boolean = false,
 
   @Schema(description = "Where a prison uses pay bands to differentiate earnings, this is the pay band given to this prisoner")
-  val prisonPayBand: PrisonPayBand,
+  var prisonPayBand: PrisonPayBand,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/Allocation.kt:31]: Not sure I understand why these have been changed to `var's`?

Comment (gerryallan-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/Allocation.kt]: Reverted to vals

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt (hunk 3 above)
Comment (gerryallan-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt:85]: Added removal of planned deallocation when end date is removed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt [hunk 5]
```diff
@@ -1,18 +1,86 @@
...
+    applyEndDateUpdate(request, allocation)
+    applyRemoveEndDateUpdate(request, allocation)
+    applyPayBandUpdate(request, allocation)
+
+    allocationRepository.saveAndFlush(allocation)
+
+    return allocation.toModel()
+  }
+
+  private fun applyStartDateUpdate(
```
[the remark this answers is not in the document]
Comment (gerryallan-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt:45]: Planned deallocation is now added when end date is added and deleted when end date is removed

**Review by matt-crowson-moj — APPROVED** [review]
LGTM - one small comment on the request object.

**Pull request merged**